docs/developers/blobstream-contracts.md

```diff
@@ -0,0 +1,149 @@
+        require(
+            blobstream.verifyAttestation(_blobstream_nonce, _tuple, _proof)
+        );
+
+        // Verify the ZKP.
+        // _tuple.dataRoot is a public input, leaves (shares) are private inputs.
+        require(verifyZKP(_rollup_block_hash, _zk_proof, _tuple.dataRoot));
+
+        // Everything checks out, append rollup block hash to list.
+        rollup_block_hashes.push(_rollup_block_hash);
+    }
+
+    function verifyZKP(
+        bytes32 _rollup_block_hash,
+        bytes calldata _zk_proof,
+        bytes32 _data_root
+    ) private pure returns (bool) {
+        return true;
+    }
+}
+```
+
+## Data structures
+
+Each [`DataRootTuple`](https://github.com/celestiaorg/blobstream-contracts/blob/master/src/DataRootTuple.sol)
+is a tuple of block height and data root. It is analogous to a Celestia block
+header. `DataRootTuple`s are relayed in batches, committed to as a
+`DataRootTuple`s root (i.e. a Merkle root of `DataRootTuple`s).
+
+The [`BinaryMerkleProof`](https://github.com/celestiaorg/blobstream-contracts/blob/master/src/lib/tree/binary/BinaryMerkleProof.sol)
+is an [RFC-6962](https://www.rfc-editor.org/rfc/rfc6962.html)-compliant Merkle
+proof. Since `DataRootTuple`s are Merkleized in a binary Merkle tree, verifying
+the inclusion of a `DataRootTuple` against a `DataRootTuple`s root requires
+verifying a Merkle inclusion proof.
+
+## Interface
+
+The [`IDAOracle`](https://github.com/celestiaorg/blobstream-contracts/blob/master/src/IDAOracle.sol)
+(**D**ata **A**vailability **O**racle Interface) interface allows L2 contracts
+on Ethereum to query the Blobstream contract for relayed `DataRootTuple`s. The
+single interface method `verifyAttestation` verifies a Merkle inclusion proof
+that a `DataRootTuple` is included under a specific batch (indexed by batch
+nonce). In other words, analogously it verifies that a specific block header is
+included in the canonical Celestia chain.
+
+## Verifying data inclusion for fraud proofs
+
+A high-level overview of how a fraud-proof based L2 would interact with
+Blobstream can be found in the [inclusion proofs documentation](https://github.com/celestiaorg/blobstream-contracts/blob/master/docs/inclusion-proofs.md).
+
+The [`DAVerifier`](https://github.com/celestiaorg/blobstream-contracts/blob/master/src/lib/verifier/DAVerifier.sol)
+library is available at `blobstream-contracts/lib/verifier/DAVerifier.sol`, and
+provides functions to verify the inclusion of individual (or multiple) shares
+against a `DataRootTuple`. The library is stateless, and assumes the
+`DataRootTuple` was previously verified as included through the Blobstream
+contract's `verifyAttestation` method.
+
+In the `DAVerifier` library, we find functions that help
+with data inclusion verification and calculating the square size of a
+Celestia block. These functions work with the Blobstream smart contract,
+using different proofs to check and confirm the data's validity. Let's
```

Comment:
```suggestion
using different proofs to check and confirm the data's validity. Let's
```
*validity* or *availability* ? we only prove that the data was posted, parsing the data and making sure it's valid is rollup specific, no?

Reply:
yes! updated this in 5f23d0011c798e3ac37f3b1e6deb3f3732ebef67